addons/sale/models/product_template.py

```diff
@@ -93,3 +145,325 @@ def get_import_templates(self):
+            if reference_product._context.get('no_variant_attribute_values'):
+                # Add "no_variant" attribute values' exclusions
+                # They are kept in the context since they are not linked to this product variant
+                parent_product_template_attribute_values |= reference_product._context.get('no_variant_attribute_values')
+
+            parent_exclusions = [
+                value_id
+                for filter_line in parent_product_template_attribute_values.mapped('exclude_for').filtered(
+                    lambda filter_line: filter_line.product_tmpl_id == self
+                ) for value_id in filter_line.value_ids.ids
+            ]
+
+        # Query all archived products for this template
+        archived_variants = self.env['product.product'].search(
+            [('product_tmpl_id', '=', self.id), ('active', '=', False)]
+        )
+
+        # TODO SEB filter them in the search?
+        if archived_variants:
+            # Old archived variants could have a different set of attributes and are not relevant here
+            # -> filter them out
+            attribute_ids = product_template_attribute_values.mapped('attribute_id')
+            archived_variants = archived_variants.filtered(
+                lambda product: all(
+                    attribute_id in product.mapped('product_template_attribute_value_ids.attribute_id')
+                    for attribute_id in attribute_ids
+                )
+            )
+
+        # array, each element is an array with the ids of an archived combination
+        archived_combinations = [archived_variant.product_template_attribute_value_ids.ids
+                for archived_variant in archived_variants]
+
+        return {
+            'exclusions': mapped_exclusions,
+            'parent_exclusions': parent_exclusions,
+            'archived_combinations': archived_combinations,
+        }
+
+    @api.multi
+    def _is_combination_possible(self, combination, reference_product=None):
+        """
+        The combination is possible if it is not excluded by any rule
+        coming from the current template, from the reference_product (if
+        given), or if there is a matching variant it should not be archived.
+
+        :param combination: recordset of product.template.attribute.value
+        :param reference_product: `product.product` from which 'self' is an
+            optional or accessory product
+        :return: True if the combination is possible else False
+        :rtype: boolean
+        """
+        self.ensure_one()
+
+        if len(combination) != len(self.attribute_line_ids):
+            # number of attribute values passed is different than the
+            # configuration of attributes on the template
+            return False
+
+        if self.attribute_line_ids.mapped('attribute_id') != combination.mapped('attribute_id'):
```

Comment:
@seb-odoo  The attribute_line_ids always contain the 'no_variant' attributes if some of those are configured on the product template.

I don't think that the provided "combination" parameter contains those 'no_variant' attributes (when used in the "create_product_variant" method for example).

This comparison does not look safe to me, could you double check?

Reply:
I changed the comment to be more clear.

It's important to always check for the full combination because if the type of an attribute has been changed, it's possible that a variant exist with no_variant attributes, which has then to be checked for possibility (if archived that would exclude it).